packages/utils/src/array.ts

```diff
@@ -0,0 +1,198 @@
+//
+// Array utilities
+//
+// Note that the array functions may be called with
+// Array-like things, such as NodeList.
+
+export const {isArray} = Array;
+
+// export function arrayForEach<T, TThis>(array: T[], action: (this: TThis, item: T, index: number, array: T[]) => void, thisArg: TThis): void;
+export function arrayForEach<T>(array: T[], action: (item: T, index: number, array: T[]) => void, thisArg?: any) {
```

Comment:
I see a lot of the same callbacks. It would be nice for DRY-ness to make a seperate type for this (even if only used internally):

    type ArrayUtilCallback<T, TResult=void> = (item: T, index: number, array: T[]) => TResult;

Reply:
@Sebazzz Where should these sorts of types be defined?

Reply:
Since it is specialized to the utility functions in this file, I'd say in this file at the top.

Reply:
Got it.

Related question:  In the case of JSX, we convert accept a number of types in `anyToNode`, with appropriate delegation by a `switch` e.g. https://github.com/knockout/tko/blob/master/packages/utils.jsx/src/JsxObserver.js#L233

I was thinking of rewriting that switch in TS with something like the following (with most specific arguments to most general):

```js
anyToNode(value: Node) : Node { ... }
anyToNode(value: null|undefined|Error|Symbol) : Node { ... }
anyToNode(value: object) : Node { ... }
anyToNode(value: function) : Node { ... }
anyToNode(value: string) : Node { ... }
anyToNode(value: boolean|number|bigint) : Node { ... }
```

If I understand it correctly, Typescript essentially is going to compile this down to a `switch` anyway.  Is there a better way to write this, or is this a sound approach? (Even this fairly naive TS version looks _**much**_ better to me, certainly, than the JS function with a switch). 

Also, the `anyToNode` returns one of `HTMComment`, `HTMLElement`, or `Text` - what's the right return type? The more general `Node`?  Or the most specific in each case? Or a combination with e.g. `HTMLElement|HTMLComment|Text`?

Reply:
I'm not familiar enough with the compiler-side of things regarding JSX, perhaps @mbest knows?
As I understand it, Typescript does not [support overloading](http://www.typescriptlang.org/docs/handbook/functions.html#overloads) in the sense that compiler figures out the checks to make, and the handbook does not express it explicitly.

Regarding the last: Yes, a union is the way to go since you know what types might be returned.

Reply:
Maybe a stupid question, why do we have these in the first place if TKO doesn't target IE8?

Reply:
@caseyWebb The `anyToNode` function is needed because it's taking user-based input and turning it into JSX; that input may be in the form of a JSX-like object (`{ elementName: string, attributes: object, children: array }`) or text, or something else that we nevertheless want to handle gracefully such as an observable or promise.

Reply:
@avickers Turns out that section of the docs is only about function type descriptions; Typescript doesn't actually do the overloading implementation.  https://github.com/Microsoft/TypeScript/issues/5658 . It should.  But it doesn't.

Reply:
That wouldn't be easy. Only for the most trivial cases overloading could be implemented.